Implement global table and table scan functionality

jadeallenx opened the pull request:
This PR implements both serial scans and parallel scans for Dynamo tables.

Commit:
Roll 0.0.5

Commit:
Support global table creation

Commit:
Add a generic message consumer for receive block

Commit:
Flesh out scanning functions

jonathn-al commented on src/darcy.erl:
```diff
@@ -10,6 +10,9 @@
 
+-define(TIMEOUT, 5000).
+-define(BIG_TIMEOUT, 60000).
+
```
Maybe more descriptive names

jadeallenx replied:
Do you have any good suggestions?

jonathn-al replied:
how about TABLE_CREATION_TIMEOUT, TABLE_SCAN_TIMEOUT or similar?

jonathn-al commented on src/darcy.erl:
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
+%%
+%% The items will be accumulated into a list and returned.  This call is
+%% syntactic sugar for `scan/4' with a function of `same(X) -> X' passed
+%% in.
```
typo? syntactic sugar for `scan_all/4'

jadeallenx replied:
Thanks. Fixed.

jonathn-al commented on src/darcy.erl:
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
+
+%% @doc This function executes a sequential table scan (no parallelism here)
+%% and returns results synchronously to the caller. If you want to scan a table
+%% with parallel workers, look at `scan/5'.
```
typo? look at `scan_parallel/5'

jonathn-al commented on src/darcy.erl:
```diff
@@ -201,6 +211,57 @@ attempt_make_table(Client, Spec) ->
+            attempt_make_global_table(Client, TableName, Regions)
+    end.
+
+global_table_setup(Client, Spec, Regions) ->
```
Not a problem, but would be neat if you included a more generic 'pmap' function (eg. similar to http://erlang.org/pipermail/erlang-questions/2009-January/041214.html) and used that here.

jadeallenx replied:
Good idea.

jonathn-al commented on src/darcy.erl:
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
+do_scan_all(_Client, _Req, {ok, empty_table}, _Fun, _Acc) -> {ok, []};
+do_scan_all(_Client, _Req, {ok, no_results}, _Fun, Acc) -> {ok, lists:reverse(lists:flatten(Acc))};
+do_scan_all(_Client, _Req, {ok, #{ <<"Items">> := I }}, Fun, Acc) ->
+    {ok, lists:reverse(lists:flatten([ lists:map(Fun, I) | Acc ]))};
```
Maybe I'm wrong, but should this be lists:flatten(lists:reverse(...)) instead?

jadeallenx replied:
Yeah, I made a function

jonathn-al commented on src/darcy.erl:
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
+
+%% @doc Scan a table in parallel using the given expression. This
+%% function is equivalent to `scan/7' with a timeout value of 60000
+%% milliseconds.
```
typo? equivalent to `scan_parallel/6'

jonathn-al commented on src/darcy.erl:
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
+        error_logger:info_msg("The scan has timed out."
+                              "Worker pids: ~p, reply pid: ~p, accumulator: ~p, errors: ~p",
+                              [Pids, Reply, Acc, Errors]),
+        Reply ! scan_timeout
```
should it be {error, scan_timeout} ?

jadeallenx replied:
hmm maybe. for now Im going to leave it as-is

jonathn-al commented on src/darcy.erl:
```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
+        ReplyPid :: pid()) -> scan_timeout |
+                              { Results :: [ term() ], Errors :: [ term() ] }.
+scan_parallel(Client, TableName, Expr, Fun, _Timeout, SegmentCount, ReplyPid) ->
+    Reqs = [ make_scan_request([Expr, table_name(TableName), make_segments(N, SegmentCount)]) ||
```
Same comment as above, might be neat if you included a general 'parallel map with up to N workers' function and used that (?)

jonathn-al reviewed (approved):
Looking good

Commit:
Clean up global table; add pmap

Clean up the global table logic.

Generalize the parallel sections into a pmap
implementation.

mrallen1 commented:
@jonathn-al Think I've implemented all of your suggestions.

jonathn-al commented:
👍
